completion/_blinky

```diff
@@ -0,0 +1,145 @@
+#compdef blinky
+# Copyright (C) 2018 Paul Seyfert <pseyfert.mathphys@gmail.com>
+#
+# This software is distributed under the terms of the GNU General Public
+# Licence version 3 (GPL Version 3), copied verbatim in the file "LICENSE".
+
+# heavily inspired by the _pacman completion
+
+# options for passing to _arguments: main pacman commands
+_blinky_opts_commands=(
+	'(-h --help)'{-h,--help}'[Display usage]'
+)
+
+_blinky_opts_common=(
+	'(-v --verbose)'{-v,--verbose}'[Be more verbose]'
+	'--local-path[Local path for building and cache]:cache_location:_files -/'
+)
+
+_blinky_keep_opts() {
+  local -a keepopts
+  keepopts=('none:keep no sources' 'skipped:keep sources of skipped packages' 'all:keep all sources')
+  _describe 'keep source setting' keepopts
+}
+
+_blinky_opts_build=(
+  '--asdeps[If packages are installed, install them as dependencies]'
+  '--force-review[Force review even if exact copies of the files have already been reviewed positively]'
+  '--keep-sources[Keep sources]:keep which sources:_blinky_keep_opts'
+  '--build-only[Only build, do not install anything]'
+)
+
+# options for passing to _arguments: options for --sync command
+_blinky_opts_sync_actions=(
+  '-S[Install package(s) from AUR]'
+  '-Sr[Reinstall package(s) from AUR (including rebuild)]'
+  '-Srr[Reinstall package(s) and dependencies from AUR (including rebuild)]'
+  '-Ss[Search for package(s) in AUR]'
+  '-Si[Get detailed info on packages in AUR]'
+  '-Syu[Upgrade all out-of-date AUR-packages]'
+  '-Sc[Clean cache of all uninstalled package files]'
+  '-Scc[Clean cache of all package files, including installed]'
+)
+
+# provides completions for packages available from repositories
+# these can be specified as either 'package' or 'repository/package'
+_blinky_completions_local_packages() {
+  local -a seq sep cmd packages repositories packages_long
+
+  packages=( $(pacman -Qm | \sed "s/ .*//") )
+
+  typeset -U packages
+  _wanted packages expl "installed non-repository packages" compadd "${(@)packages}"
+}
+
+_blinky_completions_all_packages() {
+  local -i minimum_package_name_length
+  # steer minimum package name length for launching package search
+  # set with:
+  # zstyle :completion:expand-word:complete:blinky:pkgcomp: numbers 4
```

Comment:
@cherti how do you like this. By default `blinky -S abc⇥ ` will only complete if the current package name has at least four characters (i.e. not in this example). This is threshold by the user with
```zstyle :completion:expand-word:complete:blinky:pkgcomp: numbers 4```

Reply:
ping @cherti I don't recall if we discussed this. any executive decision from your side?

The issue for me is that completion for `blinky -S <tab>` will always be on the slow side (sends a query to aur and parses the response, and even some overhead for launching a python process). otoh, not completing package names isn't great either as I would like to provide users with the functionality to just get long package names typo free by hammering tab. A threshold on a few letters avoids immediately freezing the shell by hammering tab before starting a package name (the list of all package names doesn't help a lot, one needs to type a few letters anyway. later, when there are only O(10) options, one can hammer tab or use arrows to get the right one, and can accept a second of delay easier as there's only "pick the right and hammer return" left to do, it interrupts the thought process by the user less, in my experience.)

Reply:
I do think that this looks like a viable way forward. keeping a local list of packages (and therefore having the user to actually update them) doesn't sound too appealing to me, so I think this is a very good compromise, especially as it is configurable.

Reply:
okay. suggestion where to document the configuration?

Reply:
For now the Readme, or alternatively a file `docs/completion.md`, we do not have manpages yet, this is something we probably should do at some point before 1.0.

Reply:
added to README